worlds/sc2/rules.py

```diff
@@ -877,11 +1066,45 @@ def the_reckoning_requirement(self, state: CollectionState) -> bool:
                 and (self.zerg_competent_anti_air(state)
                     or self.terran_competent_anti_air(state)
                 )
+                and self.terran_very_hard_mission_weapon_armor_upgrades(state)
+                and self.zerg_very_hard_mission_weapon_armor_upgrades(state)
             )
         else:
-            return self.zerg_competent_comp(state) and self.zerg_competent_anti_air(state)
+            return (
+                    self.zerg_competent_comp(state)
+                    and self.zerg_competent_anti_air(state)
+                    and self.zerg_very_hard_mission_weapon_armor_upgrades(state)
+            )
 
     # LotV
+    def protoss_army_weapon_armor_upgrade_count(self, state: CollectionState) -> int:
+        count: int = WEAPON_ARMOR_UPGRADE_MAX_LEVEL + 1 # +1 for Quatro
+        if self.world_has_protoss_ground_unit():
+            count = min(
+                count,
+                self.weapon_armor_upgrade_count(PROGRESSIVE_PROTOSS_GROUND_WEAPON, state),
+                self.weapon_armor_upgrade_count(PROGRESSIVE_PROTOSS_GROUND_ARMOR, state)
+            )
+        if self.world_has_protoss_air_unit():
+            count = min(
+                count,
+                self.weapon_armor_upgrade_count(PROGRESSIVE_PROTOSS_AIR_WEAPON, state),
+                self.weapon_armor_upgrade_count(PROGRESSIVE_PROTOSS_AIR_ARMOR, state)
+            )
+        if self.world_has_protoss_ground_unit() or self.world_has_protoss_air_unit():
+            count = min(
+                count,
+                self.weapon_armor_upgrade_count(PROGRESSIVE_PROTOSS_SHIELDS, state),
+            )
+        return count
```

Comment:
You add +1 for Quatro on line 1081, but don't use quatro anywhere else; recommend adding +1 if the player has quatro:

`return count + state.has(item_names.QUATRO, self.player)`

Reply:
See `weapon_armor_upgrade_count`, this counts Quatro too (so if you query just against some upgrade, it counts Quatro towards it)